### ISSUE TITLE
Outbox pattern for non-relational databases

### ISSUE THREAD

**Issue by markphillips100** [dotnet-architecture/eShopOnContainers#501]
I have read the eBook and used the code extensively to understand the pieces of the architecture.  There's one area of the architecture that puzzles me a little though in regard to the use of the outbox pattern, or lack of in all cases I should say.

I understand the business need for the outbox pattern used in the Ordering microservice to support atomically persisting domain state changes and integration events.  I'm puzzled though why the Location microservice publishes its integration events outside of its related domain state change, i.e. state can change without an event being published?

Is this a business trade-off due to the technical complexity of implementing an outbox pattern with MongoDb/CosmosDb and the business domain not really impacted by the occasional user location change event not propagating to downstream services?

What if the business were to mandate that no events can be lost.  Would this mean maybe sourcing a different db product?  But then maybe the choice of a document db was considered appropriate for other reasons in the first place.  I guess perhaps, in the case of CosmosDb, we could use a javascript stored procedure to atomically make state changes and persist an integration event.  So push the outbox pattern into the db itself?

**Comment by markphillips100** [comment]
Thanks @CESARDELATORRE for clarifying.  I guess I was trying to determine if there even is an outbox solution to nosql dbs (without 2PC), and what the guiding principles should always be when supporting eventual consistency between services using integration events.

To clarify some more, from reading up on microservices a common pattern I read in regard to integration events is the notion that they support eventual consistency.  In other words, downstream services will eventually be made aware of state changes in upstream services.  I took this to mean that eventually data is synchronized, assuming no further state changes occur.  For this statement to always be true, events produced as facts of an occurrence in a service must indeed have happened otherwise they are false events.  

It would seem to me that in order for all integration events to be true information of something that "has" happened they should only exist if the state they relate to exists also, i.e. an atomic state change.  Even though eShop doesn't do this in all cases (I realize it's not meant to be a full production system), would you say that this should be an architectural constraint for microservices, or can you think of general architectural scenarios where it's not actually necessary for eventual consistency to always be valid?

**Comment by CESARDELATORRE** [comment]
Right, microservices are eventually consistent between each other based on integration events.
In regards general architecture scenarios for making eventual consistency more or less safer with patterns like outbox, no, that really depends on what the business can asume or demands. In some business scenarios it doesn’t matter if some data is not propagated because it can be temporal data like IoT events, etc. In other business scenarios you need an outbox parttern so in case there’s any issue when sending the message you can always double check the event log for not published integration events, etc. It all depends on your business requirements. :)

**Comment by markphillips100** [comment]
Well said.  I guess if the business demand it they must be made aware of the cost associated with implementing it and whether that's a good ROI.  Not too dissimilar I guess to aggregate models in DDD becoming too large due to badly thought-out invariants.

Thanks for your thoughts @CESARDELATORRE, much appreciated.
